make_prg/recursion_tree.py

```diff
@@ -0,0 +1,404 @@
+                cluster_subalignments = NodeFactory._get_subalignments_by_clustering(alignment, clustering_result)
+                return MultiClusterNode(nesting_level, alignment, parent_node, prg_builder,
+                                        cluster_subalignments)
+
+        # here, the node is a leaf
+        return LeafNode(nesting_level, alignment, parent_node, prg_builder)
+
+    #####################################################################################################
+    #  helper methods
+    @staticmethod
+    def _alignment_has_issues(alignment: MSA) -> bool:
+        num_unique_nongapped_seqs = get_number_of_unique_ungapped_sequences(alignment)
+        too_few_unique_sequences = num_unique_nongapped_seqs <= 2
+        if too_few_unique_sequences:
+            return True
+
+        num_unique_gapped_seqs = get_number_of_unique_gapped_sequences(alignment)
+
+        # this is an assert as if it happens, something very wrong with the two previous functions is happening
+        assert num_unique_nongapped_seqs <= num_unique_gapped_seqs
+
+        alignment_has_ambiguity = num_unique_nongapped_seqs < num_unique_gapped_seqs
+        if alignment_has_ambiguity:
+            # TODO: fix alignment by deduplicating nongapped seqs and realigning them
+            # TODO: the annoying thing is that this deduplicated nongapped alignment will differ from the input alignment
+            # TODO: but I think it can be an ok solution
+            return True
+
+        return False
+
+    @staticmethod
+    def _get_nesting_level(parent_node: Optional[RecursiveTreeNode]) -> int:
+        is_root = parent_node is None
+        if is_root:
+            nesting_level = 0
+        else:
+            need_to_go_down_one_level = isinstance(parent_node, MultiClusterNode)
+            if need_to_go_down_one_level:
+                nesting_level = parent_node.nesting_level + 1
+            else:
+                nesting_level = parent_node.nesting_level
+        return nesting_level
+    #####################################################################################################
+
+    #####################################################################################################
+    #  interval methods
+    @staticmethod
+    def _infer_if_has_single_interval(all_intervals: Intervals, match_intervals: Intervals,
+                                      nesting_level: int, max_nesting: int,
+                                      alignment: MSA, min_match_length: int) -> bool:
+        single_match_interval = (len(all_intervals) == 1) and (all_intervals[0] in match_intervals)
+        if single_match_interval:
+            return True
+
+        max_nesting_level_reached = nesting_level == max_nesting
+        if max_nesting_level_reached:
+            return True
+
+        small_variant_site = alignment.get_alignment_length() < min_match_length
+        if small_variant_site:
+            return True
+
+        return False
```

Comment:
This really does not work IMO. I propose removing it and 'absorbing' it in the code proposed above

Reply:
done